ckanext/useraffiliation/actions.py

```
@@ -0,0 +1,109 @@
+import copy
+import datetime
+import logging
+from ckan import model, logic
```

Ideally, in theory, we avoid importing anything from ckan core and get everything we need from ckan.plugins.toolkit - this way they can maintain backwards compatibility between plugins and ckan core. This is a principle not always adhered to, even by CKAN core developers.  But is a good one by and large.  It's also one tip we have learn't quite recently and so you will find places in our own code which we have failed to update accordingly.  

ckan.logic isn't being used anywhere in this file, so that is easy to get rid of.  

ckan.model is only being used as a default when context['model'] is not available.  In principle, if the action has been called without context['model'] defined, it has been called wrongly. CKAN provides plenty of tools to ensure that when actions are called the model is put into the context. I think I would be tempted to just let an error raise if there is no model in the context.  The only references to "model" in the ckan plugin toolkit [explain that the model should not be imported from ckan core](https://docs.ckan.org/en/2.9/extensions/plugins-toolkit.html#ckan.plugins.toolkit.ckan.plugins.toolkit.get_action).

---

Model is also used in the create and update functions for the `if not context.get("defer_commit")` conditionals. I will be honest that this is a copy over from the example I found and I am not really sure what it's doing?

My guess is that CKAN sometimes asks the interface not to commit anything to the DB and this is catching that? I think the key thing though is that the `model.Session.commit()` inside those conditionals seems to be what adds our additional fields to the DB so is needed for this approach to work I think.

I think it's mimicking: https://github.com/ckan/ckan/blob/f58c0bcaea2184e18f2bed327873ba38c9bcbfe7/ckan/logic/action/update.py#L835

---

I have just reached the comment below explaining some of this.